homeassistant/components/integration/sensor.py

```diff
@@ -150,17 +150,29 @@ def calc_integration(event):
             old_state = event.data.get("old_state")
             new_state = event.data.get("new_state")
 
+            # We may want to update our state before an early return,
+            # based on the source sensor's unit_of_measurement
+            # or device_class.
+            update_state = False
+
             if self._unit_of_measurement is None:
                 unit = new_state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
-                self._unit_of_measurement = self._unit_template.format(
-                    "" if unit is None else unit
-                )
+                if unit is not None:
+                    self._unit_of_measurement = self._unit_template.format(
+                        "" if unit is None else unit
+                    )
+                    update_state = True
+
             if (
                 self.device_class is None
                 and new_state.attributes.get(ATTR_DEVICE_CLASS)
                 == SensorDeviceClass.POWER
             ):
                 self._attr_device_class = SensorDeviceClass.ENERGY
+                update_state = True
+
+            if update_state:
+                self.async_write_ha_state()
```

Comment:
Why not just call `self.async_write_ha_state()` ?

Reply:
I'm new to HA and don't have a good sense for how expensive async_write_ha_state() is.  The update_state = true assignment can happen in two cases above and I want to call update if one or both cases happened.  If it's very cheap to call async_write_ha_state(), I can just call it in both places above.  Would appreciate your guidance on that.

Reply:
probably you can call `self.async_write_ha_state()` regardless at the end

Reply:
There's a conditional and a potential return statement that leaves the function right below this.  I wanted to make sure it updates state before it leaves the function.
The check and return always existed, as far as I can tell.  The reason these two checks (units and device class) are above the return is because the previous author wanted to make sure the setting of units and device class happened even if the source sensor hadn't updated yet.  They just neglected to call async_wite_ha_state(), so the desired updates never actually worked.  
I see three options:
1) Leave it broken or handle that issue in a separate PR
2) Call async_write_ha_state multiple times (up to three) on execution of this function
3) Use a temp local to track whether it needs to be called (once) before the return
4) Wrap the entire method in a giant conditional so it only executes when the check for a valid state passes.  This is conceptually the same as the check/return, but introduces a level of indentation that I think is less clean.

I prefer options 2 or 3, but again, this is my first PR to HA, so very much open to your feedback.

Reply:
OK lets keep the local temp